suitcase/tiff/__init__.py

```diff
@@ -1,25 +1,312 @@
+        names `scan_XXX-'stream_name'.tiff`.
+
+        .. note::
+
+            The `stop` document is excluded as it has not been recieved yet
+            when the files are created. The `descriptor` document is excluded
+            because there is multiple 'descriptor' documents.
+
+    **kwargs : kwargs
+        kwargs to be passed to tifffile.TiffWriter.save.
+
+    Returns
+    -------
+    dest : dict
+        dict mapping the 'labels' to the file names
+    """
+
+    serializer = Serializer(directory, file_prefix, **kwargs)
+    try:
+        for item in gen:
+            serializer(*item)
+    finally:
+        serializer.close()
+
+    return serializer.artifacts
+
+
+class Serializer(event_model.DocumentRouter):
+    """ Serialize a set of (name, document) tuples to tiff format(s) and one a
+    JSON format for the metadata.
+
+    The structure of the json is:
+
+            {'metadata': {'start': start_doc, 'stop': stop_doc,
+                          'descriptors': {stream_name1: 'descriptor',
+                                          stream_name2: ...}},
+             stream_name1: {'seq_num': [], 'uid': [], 'time': [],
+                         'timestamps': {det_name1:[], det_name2:[],...},
+             stream_name2: ...}}
+
+            .. note::
+
+                This schema was chosen as the API is very similar to the
+                intake-databroker API. The same schema is used for all json
+                files created with our base suitcase export functions.
+
+    Parameters
+    ----------
+    gen : generator
+        expected to yield (name, document) pairs
+
+    directory : string, Path or Wrapper
+        The filepath and filename suffix to use in the output files or a file
+        handle factory wrapper(see ADD LINK HERE). An empty string will place
+        the file in the current directory.
+
+    file_prefix : str
+        An optional prefix for the file names that will be created, default is
+        an empty string.A templated string may also be used, where curly
+        brackets will be filled in with the attributes of the 'start'
+        documents.
+        e.g., `file_prefix`="scan_{start[scan_id]}-" will result in files with
+        names `scan_XXX-'stream_name'.tiff`.
+
+        .. note::
+
+            The `stop` document is excluded as it has not been recieved yet
+            when the files are created. The `descriptor` document is excluded
+            because there is multiple 'descriptor' documents.
+
+    **kwargs : kwargs
+        kwargs to be passed to tifffile.TiffWriter.save.
+
+    .. note::
+
+        It is the resonsibility of whatever creates this class to close the
+        used file handles when done. To do this use the lines below when
+        everything is complete:
+
+        .. code::
+            for artifact in serializer.artifacts:
+                artifacts.close()
+    """
+
+    def __init__(self, directory, file_prefix, **kwargs):
```

Comment:
The same argument for a non-empty default `file_prefix` applies in this layer, I think. It would be nice if this just worked and did not lead to overwite errors. Again, further customization will be possible and encouraged, but the low-effort case will at least _work_.

```py
def serializer_factory(start_doc):
    serializer = Serializer(DIRECTORY)
    serializer.start(start_doc)
    return [serializer], []

RE.subscribe(RunRouter([serializer_factory]))
```

Reply:
thanks for the reminder this is now done